src/test/java/io/supertokens/test/totp/TOTPStorageTest.java

```diff
@@ -0,0 +1,465 @@
+
+    @AfterClass
+    public static void afterTesting() {
+        Utils.afterTesting();
+    }
+
+    @Before
+    public void beforeEach() {
+        Utils.reset();
+    }
+
+    public TestSetupResult initSteps() throws InterruptedException {
+        String[] args = { "../" };
+
+        TestingProcessManager.TestingProcess process = TestingProcessManager.start(args);
+        assertNotNull(process.checkOrWaitForEvent(ProcessState.PROCESS_STATE.STARTED));
+
+        if (StorageLayer.getStorage(process.getProcess()).getType() != STORAGE_TYPE.SQL) {
+            assert (false);
+        }
+        TOTPSQLStorage storage = StorageLayer.getTOTPStorage(process.getProcess());
+
+        return new TestSetupResult(storage, process);
+    }
+
+    private static TOTPUsedCode[] getAllUsedCodesUtil(TOTPStorage storage, String userId)
+            throws StorageQueryException, StorageTransactionLogicException {
+        assert storage instanceof TOTPSQLStorage;
+        TOTPSQLStorage sqlStorage = (TOTPSQLStorage) storage;
+
+        return (TOTPUsedCode[]) sqlStorage.startTransaction(con -> {
+            TOTPUsedCode[] usedCodes = sqlStorage.getAllUsedCodesDescOrder_Transaction(con, userId);
+            sqlStorage.commitTransaction(con);
+            return usedCodes;
+        });
+    }
+
+    private static void insertUsedCodesUtil(TOTPSQLStorage storage, TOTPUsedCode[] usedCodes)
+            throws StorageQueryException, StorageTransactionLogicException, TotpNotEnabledException,
+            UsedCodeAlreadyExistsException {
+        try {
+            storage.startTransaction(con -> {
+                try {
+                    for (TOTPUsedCode usedCode : usedCodes) {
+                        storage.insertUsedCode_Transaction(con, usedCode);
+                    }
+                } catch (TotpNotEnabledException | UsedCodeAlreadyExistsException e) {
+                    throw new StorageTransactionLogicException(e);
+                }
+                storage.commitTransaction(con);
+
+                return null;
+            });
+        } catch (StorageTransactionLogicException e) {
+            Exception actual = e.actualException;
+            if (actual instanceof TotpNotEnabledException) {
+                throw (TotpNotEnabledException) actual;
+            } else if (actual instanceof UsedCodeAlreadyExistsException) {
+                throw (UsedCodeAlreadyExistsException) actual;
+            }
+            throw new StorageQueryException(e);
+        }
+    }
+
+    @Test
+    public void createDeviceTests() throws Exception {
+        TestSetupResult result = initSteps();
+        TOTPSQLStorage storage = result.storage;
+
+        TOTPDevice device1 = new TOTPDevice("user", "d1", "secret", 30, 1, false);
+        TOTPDevice device2 = new TOTPDevice("user", "d2", "secret", 30, 1, true);
+        TOTPDevice device2Duplicate = new TOTPDevice("user", "d2", "new-secret", 30, 1, false);
+
+        storage.createDevice(device1);
```

Comment:
where are tests in which this function throws different errors and you catch them and check if the right error is thrown?

Reply:
Let's discuss all tests you have done over a call..